lib/Transforms/IPO/StructAnalysisCloseProximity.cpp

```diff
@@ -0,0 +1,1004 @@
+              dbgs() << "Connect new node with previous node in the BB: Field ["
+                     << BBI->LastNode->FieldNum << "] to ["
+                     << FieldNum.getValue() << "] with count: " << C
+                     << " and data " << D << " bytes \n");
+          BBI->LastNode = NewNode;
+        } else {
+          // If there's no previous node, update BBI info
+          DEBUG_WITH_TYPE(
+              DEBUG_TYPE_FRG,
+              dbgs() << "No previous node found. It is the first node\n");
+          BBI->FirstNode = BBI->LastNode = NewNode;
+        }
+      } else {
+        // Case that I is not struct access but a memory access
+        if (I.getOpcode() == Instruction::Load ||
+            I.getOpcode() == Instruction::Store) {
+          DEBUG_WITH_TYPE(DEBUG_TYPE_FRG,
+                          dbgs()
+                              << "Found an instruction " << I
+                              << " is not a struct access but a load/store.\n");
+          if (BBI->LastNode == NULL) {
+            // Create a dummy node for the first non-struct memory access
+            DEBUG_WITH_TYPE(
+                DEBUG_TYPE_FRG,
+                dbgs() << "Create a dummy node as the first node in the BB.\n");
+            BBI->FirstNode = BBI->LastNode = FRG->createNewNode();
+          } else {
+            // Record the bytes accessed, used to calculate distance
+            BBI->RemainBytes += getMemAccessDataSize(&I);
+            DEBUG_WITH_TYPE(DEBUG_TYPE_FRG,
+                            dbgs() << "Increment remaining data size to "
+                                   << BBI->RemainBytes << "\n");
+          }
+        }
+      }
+    }
+    if (BBI->LastNode == NULL) {
+      // If all instructions are examined and the BB doesn't have any memory
+      // accesses, create a dummy node
+      assert(BBI->FirstNode == NULL);
+      DEBUG_WITH_TYPE(
+          DEBUG_TYPE_FRG,
+          dbgs()
+              << "Create a dummy node as BB does not have memory accesses\n");
+      BBI->FirstNode = BBI->LastNode = FRG->createNewNode();
+    }
+  }
+  // Before connect nodes, need to mark all back edges first
+  detectBackEdges(FRG, F);
+  // Connect nodes between different basic blocks
+  for (auto &BB : *F) {
+    auto *BBI = FRG->getBasicBlockHelperInfo(&BB);
+    auto *Term = BB.getTerminator();
+    // Check all the connections from BB in CFG
+    for (const auto *SB : Term->successors()) {
+      auto *SBI = FRG->getBasicBlockHelperInfo(SB);
+      ExecutionCountType C = 0;
+      auto BBCount = StructInfo->getExecutionCount(&BB);
+      auto SBCount = StructInfo->getExecutionCount(SB);
+      if (BBCount.hasValue() && SBCount.hasValue()) {
+        // Take probability of the branch
+        auto Prob = StructManager->getBranchProbability(&BB, SB);
+        assert(Prob.hasValue());
+        C = std::min(BBCount.getValue() * Prob.getValue(),
+                     SBCount.getValue() * 1.0);
```

Comment:
I was seeing a case where succ has a lower count before we changed the BlockFrequencyInfo. I don't know if it still matters now... Can I just keep it this way and test if we can safely remove it later? Also, multiply by 1.0 is used to convert it to double... Should I use cast<>() instead?